jena-text/src/main/java/org/apache/jena/query/text/TextIndexES.java

```diff
@@ -0,0 +1,394 @@
+     */
+    @Override
+    public void rollback() {
+       //Do Nothing
+
+    }
+
+    /**
+     * We don't have resources that need to be closed explicitely
+     */
+    @Override
+    public void close() {
+        // Do Nothing
+
+    }
+
+    /**
+     * Update an Entity. Since we are doing Upserts in add entity anyways, we simply call {@link #addEntity(Entity)}
+     * method that takes care of updating the Entity as well.
+     * @param entity the entity to update.
+     */
+    @Override
+    public void updateEntity(Entity entity) {
+        //Since Add entity also updates the indexed document in case it already exists,
+        // we can simply call the addEntity from here.
+        addEntity(entity);
+    }
+
+
+    /**
+     * Add an Entity to the ElasticSearch Index.
+     * The entity will be added as a new document in ES, if it does not already exists.
+     * If the Entity exists, then the entity will simply be updated.
+     * The entity will never be replaced.
+     * @param entity the entity to add
+     */
+    @Override
+    public void addEntity(Entity entity) {
+        LOGGER.debug("Adding/Updating the entity in ES");
+
+        //The field that has a not null value in the current Entity instance.
+        //Required, mainly for building a script for the update command.
+        String fieldToAdd = null;
+        String fieldValueToAdd = "";
+        try {
+            XContentBuilder builder = jsonBuilder()
+                    .startObject();
+
+            for(String field: docDef.fields()) {
+                if(entity.get(field) != null) {
+                    if(entity.getLanguage() != null && !entity.getLanguage().isEmpty() && isMultilingual) {
+                        fieldToAdd = field + "_" + entity.getLanguage();
+                    } else {
+                        fieldToAdd = field;
+                    }
+
+                    fieldValueToAdd = (String) entity.get(field);
+                    builder = builder.field(fieldToAdd, Arrays.asList(fieldValueToAdd));
+                    break;
+                } else {
+                    //We are making sure that the field is at-least added to the index.
+                    //This will help us tremendously when we are appending the data later in an already indexed document.
+                    builder = builder.field(field, Collections.emptyList());
+                }
+
+            }
+
+            builder = builder.endObject();
+            IndexRequest indexRequest = new IndexRequest(indexName, docDef.getEntityField(), entity.getId())
+                    .source(builder);
+
+            String addUpdateScript = "if(ctx._source.<fieldName> == null || ctx._source.<fieldName>.empty) " +
+                    "{ctx._source.<fieldName>=['<fieldValue>'] } else {ctx._source.<fieldName>.add('<fieldValue>')}";
+            addUpdateScript = addUpdateScript.replaceAll("<fieldName>", fieldToAdd).replaceAll("<fieldValue>", fieldValueToAdd);
```

Comment:
Reworked the scripts to pass fieldValue as parameters. This will result in the script NOT being recompiled every time.
Can not pass the field name as parameter because apparently ES does a param check with the values (right side of = sign).
In any case this would not result in any extra performance degradation because the script is compilable per field type.
Also if the value contains single quote then the single quote is preserved. Tested it with JenaESTextExample and the latest scripting update.